scripts/perfcontrib.py

```diff
@@ -0,0 +1,116 @@
+from json import loads, dumps
+from urllib.request import urlopen, Request
+from urllib.parse import urlencode
+from urllib.error import HTTPError
+import time
+import sys
+import os
+
+tenantId = "72f988bf-86f1-41af-91ab-2d7cd011db47"
+appId = "c2fe4cd0-be4a-468b-aa4f-078c67dcab6e"
+
+uploadService = "https://perfcontrib.azurewebsites.net"
+uploadEndpoint = f"{uploadService}/api/UploadPerfData"
+authEndpoint = f"{uploadService}/.auth/login/aad"
+authDetailsEndpoint = f"{uploadService}/.auth/me"
+
+aadUrl = f"https://login.microsoftonline.com/{tenantId}"
+
+def get_token() -> str:
+    path = os.path.expanduser("~/.perfcontrib")
+
+    token: str | None
+
+    try:
+        if not os.path.exists(path):
+            os.makedirs(path)
+        with open(os.path.expanduser("~/.perfcontrib/token")) as tokenfile:
+            token = tokenfile.readline()
+    except FileNotFoundError:
+        pass
+
+    if token:
+        try:
+            with urlopen(Request(authDetailsEndpoint,
+                    headers = { "X-ZUMO-AUTH": token })) as response:
+                print("Using cached credentials.")
+        except HTTPError as error:
+            token = None
+
+    if not token:
+        token = authenticate()
+    
+    if token:
+        with open(os.path.expanduser("~/.perfcontrib/token"), "w") as tokenfile:
```

Comment:
I believe best practice is to never let tokens touch the disk (?) ... is this token expected to be public? In which case, protecting it obviously doesn't matter.

Reply:
That's a good bias; here are some thoughts regarding this specific case. This token isn't particularly sensitive (see **Background** below). Not storing it at all means that perf friends will have to log in to AAD using the **Device Token flow** (see below) for each monthly perf run they wish to contribute.

If we're willing to complicate `benchmarks-monthly.py` to require installation of python dependencies, `msal` and `msal-extensions` (which is what the Azure CLI uses for its similar login flow) has an implementation of cross-platform data encryption for its token store (DPAPI on Windows, KeyChain on macOS, etc).

A wilder option, I suppose, would be to find a way to persist this token in an in-memory service for the duration of a couple of perf data contributions.

For what it's worth, users will have to log in each month on each device from which they're contributing data, so it's entirely possible that this is the only exceptional month where most people are uploading two or more datasets for which this token cache would have any impact.

**Background**
The token in question is a user- (and device-) specific jwt that authenticates you to only the perfcontrib endpoint for 30 days. It's issued and managed by Azure Funtions ASP.NET-style EasyAuth (in this case, kinda like a session cookie, but for a CLI).

Here's an example of the complete claims that it contains:
```json
{
  "stable_sid": "sid:153afaa8a107203cd2eb9d6aeb2f40ca",
  "sub": "sid:4e1d5eff56c861acbad180778084720d",
  "idp": "aad",
  "ver": "3",
  "nbf": 1652119044,
  "exp": 1654711044,
  "iat": 1652119044,
  "iss": "https://perfcontrib.azurewebsites.net/",
  "aud": "https://perfcontrib.azurewebsites.net/"
}
```

**Device Token flow**
![image](https://user-images.githubusercontent.com/31878/167509765-f7d1f61b-9b66-4b58-a20a-946000695bc3.png)

Reply:
> Is it possible to shorten the validity time that these tokens are valid when creating them?

Perhaps. I'll investigate that on the Azure Functions EasyAuth side of things.